### ISSUE TITLE
write_fmt instead of write_str

### ISSUE THREAD

**Issue by Yuri6037** [rust-lang/rust-clippy#7941]
### What it does

I propose to add a lint to detect uses of `write_str(format!)` and suggest to replace them by `write_fmt(format_args!)`.

This would be usefull in all implementation of Display.

This proposal is motivated from one of my first projects where I've implemented Display for all of my error types using write_str where I could have used write_fmt. I've seen in some other project of mine where I've used write_fmt where applicable I could have a bit better performance on my CPU.

### Categories (optional)

- Kind: perf, maybe as well complexity

*What is the advantage of the recommended code over the original code*

- format! has to allocate a String anyway to store the formatted string inside.
- The proposed code is less complicated (no need to take reference to formatted string can just pass the macro directly).

### Drawbacks

- Longer to type.

### Example

```rust
fmt.write_str(&format!("Some string with pars {} {}", par1, par2))
```

Could be written as:

```rust
fmt.write_fmt(format_args!("Some string with pars {} {}", par1, par2))
```

**Comment by xfix** [comment]
This has a drawback of being slower for unbuffered writers. That said, writing text data into an unbuffered writer is almost certainly a bad idea, so this is more of an edge case.

**Comment by Yuri6037** [comment]
> This has a drawback of being slower for unbuffered writers (in particular `File` and `TcpStream`). That said, writing text data into an unbuffered writer is almost certainly a bad idea, so this is more of an edge case that realistically doesn't have to be considered by this lint.

Yeah probably that might be slower on File and TcpStream. It's true that I didn't think of this case as I intend this lint primarily for Display on error types. Usually you just print errors to standard error in some cases you might want to log. Even in the case of logging most loggers would rely on buffered backends. I hardly see a case you'd on purpose use an unbuffered writer to write error type. Even in case of networking you want most of the time a smaller buffer to store the packet you're currently building.

**Comment by xfix** [comment]
Definitely agreeing with you here :). As I said, "writing text data into an unbuffered writer is almost certainly a bad idea". It's just something that probably should be documented in lint implementation.
